geofront/server.py

```diff
@@ -1033,6 +1042,58 @@ def authorize_remote(token_id: str, alias: str):
     )
 
 
+@sockets.route('/ws/tokens/<token_id:token_id>/remotes/<alias>/ssh/')
+def proxy_ssh(ws, token_id: str, alias: str):
+    team = get_team()
+    identity = get_identity(token_id)
+    key_store = get_key_store()
+    master_key_store = get_master_key_store()
+    remotes = get_remote_set()
+    policy = get_permission_policy()
+    try:
+        remote = remotes[alias]
+    except KeyError:
+        response = jsonify(
+            error='not-found',
+            message='No such remote alias: {}.'.format(alias)
+        )
+        response.status_code = 404
+        return response
+    if not policy.permit(remote, identity, team.list_groups(identity)):
+        response = jsonify(
+            error='forbidden',
+            message='Remote {0} is disallowed.  Contact the system '
+                    'administrator.'.format(alias)
+        )
+        response.status_code = 403
+        return response
+
+    ssh_sock = create_connection((remote.host, remote.port))
+
+    def pipe():
+        while True:
+            try:
+                reply = ssh_sock.recv(4096)
+            except OSError:
```

Comment:
It is to handle [this](https://github.com/gevent/gevent/blob/f3acb176d0f0f1ac797b50e44a5e03726f687c53/src/gevent/_fileobjectcommon.py#L11-L13) and there is no intrinsic Exception classes corresponding to EBADF.